designs/autoscaler.md

```diff
@@ -0,0 +1,526 @@
+* Integrate autoscaling configuration in cluster spec to make EKS-A cluster autoscaling-aware.
+* Generate an EKS-A opinionated Cluster Autoscaler curated package template with default configurations.
+* Install Cluster Autoscaler controller as a curated package during or after management/workload cluster creation.
+* Support upstream k8s Autoscaler installed/managed outside of EKS-A Curated Package.
+* Update the autoscaling configuration to set up the min/max number of nodes in each worker node group.
+
+#### Out of Scope
+
+* Compare cluster autoscaler products (e.g. [Karpenter](https://karpenter.sh/))
+* Autoscale at the pod level
+* Control Plane nodes autoscaling
+
+#### Future Scope
+
+* Support other autoscaling solutions (e.g. [Karpenter](https://karpenter.sh/)).
+
+## Overview of Solution
+
+We provide a solution to add autoscaling configuration in the EKS Anywhere API, which marks the worker node groups to be Cluster Autoscaler compatible. EKS Anywhere user can then install Cluster Autoscaler in any EKS-A cluster with proper configuration, and expect node autoscaling to work for the associated resource group.
+
+1. Introduce an optional type `AutoScalingConfiguration` in the cluster and populate it into each worker node group configuration.
+2. If `AutoScalingConfiguration` is specified (not nil) for a worker node group, EKS-A converts its fields to CAPI autoscaler annotations and applies them to the underlying CAPI Machine Deployment resource. In this way, the specified worker node group will be discoverable by the Cluster Autoscaler controller once installed.
+3. Install Cluster Autoscaler either (A) as an EKS-A curated package, or (B) from upstream; with configuration that uses Cluster API as cloud provider, and auto-discovers node groups in the target cluster.
+
+### API Design
+
+```go
+type WorkerNodeGroupConfiguration struct {
+    // Count defines the number of desired worker nodes.
+    Count *int `json:"count,omitempty"`
+    
+    // AutoScalingConfiguration defines the auto scaling configuration
+    // +optional
+    AutoScalingConfiguration *AutoScalingConfiguration `json:"autoscalingConfiguration,omitempty"`
+    ...
+}
+
+type AutoScalingConfiguration struct {
+    // MinCount defines the minimum number of nodes for the associated resource group.
+    // +optional
+    MinCount int `json:"minCount,omitempty"`
+    
+    // MaxCount defines the maximum number of nodes for the associated resource group.
+    // +optional
+    MaxCount int `json:"maxCount,omitempty"`
+}
+```
+
+### `AutoScalingConfiguration` Defaults and Validation
+
+We consider the following semantics when setting initial Count that is passed down to MachineDeployment:
+
+* If `Count` is specified, regardless the existence of `AutoScalingconfiguration`, `Count` is used as the initial count.
+* If `AutoScalingconfiguration` is specified, then `Count` is optional. When `Count` is not specified: `MinCount` is used as the initial count.
```

Comment:
Having this documented is great. I can see it helping give us a defined path to take in the event that someone configures for an autoscaler, but doesn't have paid curated packages support. This way you've designed in a clearly defined and workable fallback strategy. 👍🏻